### PR TITLE
Add Google OAuth support

### PR DESCRIPTION
**Pull request by ChristophWurst**
Clean version of https://github.com/nextcloud/mail/pull/6830. Reviewable commit by commit.

Fixes https://github.com/nextcloud/mail/issues/6454

## To do

- [x] Google Oauth admin settings
- [x] Account setup flow adaptation
  - [x] Requires https://github.com/nextcloud/nextcloud-vue/pull/3369
- [x] Token refresh logic
- [x] Account re-auth for expired/revoke refresh tokens
  - [x] https://github.com/nextcloud/mail/pull/7476
- [x] (App) password to oauth migration
- [x] Edge case: email mismatch between Mail form and Google consent flow. ~~Fetch email from Gmail so the token definitely works.~~ Consent windows allows us to "hint" the account email address
- [x] Documentation

### Scenario one: fresh ~~avocado~~ account
1) Follow the admin docs to obtain and save a client ID and secret
2) Open the app
3) Enter your gmail address
  * See the form submit button change to *Sign in with Google*
4) Follow the instructions on the screen

### Scenario two: account migration
0) Unlink gmail oauth if you set a client ID/secret already
1) Enable 2FA for your gmail account, create an app password
2) Open the app
3) Add the account using the app password - the form submit button should say *Connect*
4) Follow the admin docs to obtain and save a client ID and secret
5) Open the app - account should still sync and open mail
6) Open account settings, scroll to server settings - the form submit button should say 'Reconnect Google account*
7) Reconnect the account - it now uses oauth instead of the app password

## Screenshots

| Screen | Shot |
|---|---|
| Admin settings | Bildschirmfoto vom 2022-10-14 13-42-14 |
| Before setup | Bildschirmfoto vom 2022-10-14 17-03-37 |
| During setup | Bildschirmfoto vom 2022-10-14 17-10-37 |

### CODE REVIEW

**Comment by ChristophWurst** [comment]
https://github.com/nextcloud/nextcloud-vue/releases/tag/v7.1.0-beta.0 has arrived

**Comment by miaulalala** [comment]
Right, need to setup my own NC to do this. Weekend plans, set!!! Noice!

**Comment by miaulalala** [comment]
> You can test this on your dev instance :)

I don't think so, I need a valid domain but I only have the setup locally for the consent screen.

**Comment by ChristophWurst** [comment]
Bildschirmfoto vom 2022-11-21 09-45-01

This works. I will change the admins settings to print that URL for the admins.

File: lib/Integration/GoogleIntegration.php [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		$clientSecret = $this->crypto->decrypt($encryptedClientSecret);
+		$httpClient = $this->clientService->newClient();
+		try {
+			$response = $httpClient->post('https://oauth2.googleapis.com/token', [
+				'content-type' => 'application/json',
+				'body' => json_encode([
+					'client_id' => $clientId,
+					'client_secret' => $clientSecret,
+					'grant_type' => 'authorization_code',
+					'redirect_uri' => 'https://localhost/apps/mail/integration/google-auth', // TODO: fix
```
Comment (ChristophWurst) [inline lib/Integration/GoogleIntegration.php:113]: Whoops

**Comment by ChristophWurst** [comment]
Bildschirmfoto vom 2022-11-21 10-03-30

**Comment by miaulalala** [comment]
What I'm trying to say is I can't test it bc my google consent screen configuration won't let me use:

localhost (as this is not my request domain)
nor 
nextcloud.local (as this is not a valid TLD)

**Comment by ChristophWurst** [comment]
Did you add your test email account to the list of allowed email addresses for the unverified oauth client? This might block the consent request as well.

**Comment by ChristophWurst** [comment]
Bildschirmfoto vom 2022-11-21 14-44-41

**Comment by miaulalala** [comment]
Error is the redirect uri:

The app wants to use:
image

But my config is:
image

I am not sure if that is an issue of your PR or an issue of my OAuth credentials. I'm assuming the latter but please confirm.

**Comment by ChristophWurst** [comment]
What URL does Nextcloud generate for you at https://github.com/nextcloud/mail/pull/7430#issuecomment-1321724723?

**Comment by ChristophWurst** [comment]
Nextcloud will use the Nextcloud request host for the absolute URL. So please either access the app at *localhost* or use ``'overwrite.cli.url' => 'https://localhost',`` in your config.php to force localhost

**Comment by ChristophWurst** [comment]
Guess what, I forgot to add XOAUTH2 to the SMTP client factory

**Comment by GretaD** [comment]
here its how mailboxes look for me.
Screenshot from 2022-11-24 13-58-52

Things i noticed:

- Mailbox named _all_ shows Inbox and sent msg.
- Gmail mailbox loads forever
- Sending/received emails works well, but messages sent to my gmail account are always shown as a thread of 2 envelops(same envelope shown twice)
- My gmail account was shown twice, this can be because i tried yesterday to log in but couldnt because of google authentication and the account was listed as disabled.

I had to hard refresh couple of times and load all mailboxes at least once before i was able to do anything like sending or receiving msg. First try they load forever, then i refreshed again and empty content was shown then all was fine and envelops start showing.

**Comment by ChristophWurst** [comment]
> Gmail mailbox loads forever

Errors? Logs?

**Comment by GretaD** [comment]
> > Gmail mailbox loads forever
> 
> Errors? Logs?

yes, sorry, i updated my comment

**Comment by ChristophWurst** [comment]
> * Gmail mailbox loads forever

Are you talking about `[Gmail]` literally or any mailbox of your gmail account?

**Comment by ChristophWurst** [comment]
`[Gmail]` doesn't work with password auth either. It must be a permission problem or something else.

**Comment by miaulalala** [comment]
Tested it again to get to the bottom of the [GMAIL] issue and found another issue. When removing the gmail account, it stays authorised and can't be added again. It might be an issue with the popup not passing on the "already authorised" response properly:

Peek 2022-11-24 22-36

**Comment by miaulalala** [comment]
Looks like we're not filtering the right attributes. `\noselect` is done, but not `\\noselect`. (Please mind that I had to edit the markdown to display two backslashes :) )

|id |name                         |account_id|attributes                                   |delimiter|messages|unseen|selectable|special_use|sync_new_lock|sync_changed_lock|sync_vanished_lock|sync_new_token              |sync_changed_token          |sync_vanished_token         |sync_in_background|
|---|-----------------------------|----------|---------------------------------------------|---------|--------|------|----------|-----------|-------------|-----------------|------------------|----------------------------|----------------------------|----------------------------|------------------|
|527|[Gmail]/Starred              |50        |["\\\flagged","\\\hasnochildren","\\\subscribed"]|/        |0       |0     |1         |["flagged"]|             |                 |                  |                            |                            |                            |0                 |
|526|[Gmail]/Spam                 |50        |["\\\hasnochildren","\\\junk","\\\subscribed"]  |/        |9       |9     |1         |["junk"]   |             |                 |                  |                            |                            |                            |0                 |
|525|[Gmail]/Sent Mail            |50        |["\\\hasnochildren","\\\sent","\\\subscribed"]  |/        |139     |0     |1         |["sent"]   |             |                 |                  |                            |                            |                            |0                 |
|524|[Gmail]/Important            |50        |["\\\hasnochildren","\\\important","\\\subscribed"]|/        |1       |0     |1         |[]         |             |                 |                  |                            |                            |                            |0                 |
|523|[Gmail]/Drafts               |50        |["\\\drafts","\\\hasnochildren","\\\subscribed"]|/        |0       |0     |1         |["drafts"] |             |                 |                  |                            |                            |                            |0                 |
|522|[Gmail]/Bin                  |50        |["\\\hasnochildren","\\\subscribed","\\\trash"] |/        |206     |199   |1         |["trash"]  |             |                 |                  |                            |                            |                            |0                 |
|521|[Gmail]/All Mail             |50        |["\\\all","\\\hasnochildren","\\\subscribed"]   |/        |720     |312   |1         |["all"]    |             |                 |                  |VTEwNDE4LFYxMSxIMTMyNzMxMQ==|VTEwNDE4LFYxMSxIMTMyNzIzNw==|VTEwNDE4LFYxMSxIMTMyNzIzNw==|0                 |
|520|[Gmail]                      |50        |["\\\hashchildren","\\\nonexistent","\\\subscribed","\\\noselect"]|/        |0       |0     |0         |[]         |             |                 |                  |                            |                            |                            |0                 |
|519|Test                         |50        |["\\\hasnochildren","\\\subscribed"]           |/        |0       |0     |1         |[]         |             |                 |                  |                            |                            |                            |0                 |
|518|Notes                        |50        |["\\\hasnochildren","\\\subscribed"]           |/        |1       |0     |1         |[]         |             |                 |                  |                            |                            |                            |0                 |
|517|INBOX                        |50        |["\\\hasnochildren","\\\subscribed"]           |/        |576     |312   |1         |["inbox"]  |             |                 |                  |VTY5MzYsVjEsSDEzMjczMTE=    |VTY5MzYsVjEsSDEzMjczMDc=    |VTY5MzYsVjEsSDEzMjczMDc=    |0                 |
|516|Google                       |50        |["\\\hasnochildren","\\\subscribed"]           |/        |0       |0     |1         |[]         |             |                 |                  |                            |                            |                            |0                 |

**Comment by miaulalala** [comment]
The mails exist twice, that's the duplication issue:

|id |uid                          |message_id|mailbox_id                                   |subject|sent_at|flag_answered|flag_deleted|flag_draft |flag_flagged|flag_seen|flag_forwarded|flag_junk                   |flag_notjunk                |flag_attachments            |flag_important|structure_analyzed|preview_text                                                                                                                                                                                                                                                   |updated_at|references|in_reply_to|thread_root_id                                                                    |flag_mdnsent|imip_message|imip_processed|imip_error|
|---|-----------------------------|----------|---------------------------------------------|-------|-------|-------------|------------|-----------|------------|---------|--------------|----------------------------|----------------------------|----------------------------|--------------|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|----------|-----------|----------------------------------------------------------------------------------|------------|------------|--------------|----------|
|56538|10417                        |`<01000184ab2ff3c3-8ad04bf7-6dd8-4743-849d-c1edcd3f097c-000000@email.amazonses.com>`|521                                          |Failbetter Games News November 2022: Fonts that tickle your fancy|1669319357|0            |0           |0          |0           |1        |0             |0                           |0                           |0                           |0             |1                 |Hello, delicious friend!  A juicy newsletter for you today. Let’s feast.  FALLEN LONDON  _The island is a node. You feel the connections spiralling out, everything hitched to every other thing as your view widens to encompass an entire universe. _  _Th   |1669329081|[]        |           |`<01000184ab2ff3c3-8ad04bf7-6dd8-4743-849d-c1edcd3f097c-000000@email.amazonses.com>`|0           |0           |0             |0         |
|55818|6935                         |`<01000184ab2ff3c3-8ad04bf7-6dd8-4743-849d-c1edcd3f097c-000000@email.amazonses.com>`|517                                          |Failbetter Games News November 2022: Fonts that tickle your fancy|1669319357|0            |0           |0          |0           |1        |0             |0                           |0                           |0                           |0             |1                 |Hello, delicious friend!  A juicy newsletter for you today. Let’s feast.  FALLEN LONDON  _The island is a node. You feel the connections spiralling out, everything hitched to every other thing as your view widens to encompass an entire universe. _  _Th   |1669326604|[]        |           |`<01000184ab2ff3c3-8ad04bf7-6dd8-4743-849d-c1edcd3f097c-000000@email.amazonses.com>`|0           |0           |0             |0         |

Once in INBOX, once in All

**Comment by miaulalala** [comment]
Tracked it down to the linked PR.

**Comment by miaulalala** [comment]
WHat I also found was when adding an account, removing it and then adding it again, the signup process isn't smooth any more. i don't know why that is but I had to do the login twice any time the app is already authorised with Gmail to get to a logged in state

**Comment by ChristophWurst** [comment]
> I had to do the login twice any time the app is already authorised with Gmail to get to a logged in state

The login into your Google account?

**Comment by miaulalala** [comment]
> > I had to do the login twice any time the app is already authorised with Gmail to get to a logged in state
> 
> The login into your Google account?

yes. So the flow would be: 
- enter gmail address
- get oauth popup
- accept the popup
- popup closes but login form is displayed with the popup closed text
- do the whole thing again
- finally arrive at the logged in mail account page

**Comment by ChristophWurst** [comment]
> popup closes but login form is displayed with the popup closed text

When my popup closes I see

Bildschirmfoto vom 2022-12-01 15-35-03

and then

Bildschirmfoto vom 2022-12-01 15-35-10

No double login.

File: src/components/AccountForm.vue [hunk 2]
```diff
@@ -529,6 +606,8 @@ export default {
 						this.feedback = t('mail', 'IMAP connection failed')
 					} else if (error.data?.service === 'SMTP') {
 						this.feedback = t('mail', 'SMTP connection failed')
+					} else if (error.message === CONSENT_ABORTED) {
+						this.feedback = t('mail', 'Google authorization popup closed')
```
Comment (ChristophWurst) [inline src/components/AccountForm.vue:610]: @miaulalala is that the message you see?

Comment (ChristophWurst) [reply src/components/AccountForm.vue:610]: That could mean your popup window can't post the message to the parent window and just closes. Then the parent window wrongly assumes the popup was closed prematurely and handles it as aborted consent. I'll investigate.

Comment (ChristophWurst) [reply src/components/AccountForm.vue:610]: Same behavior with FF and Chrome.

Comment (miaulalala) [reply src/components/AccountForm.vue:610]: yes

Comment (miaulalala) [reply src/components/AccountForm.vue:610]: was a caching issue on the oauth consent window

**Comment by ChristophWurst** [comment]
Alright. This works. Could I get a review and approval @GretaD @miaulalala? :)

File: doc/admin.md [hunk 3]
```diff
@@ -68,6 +68,13 @@ occ config:app:set mail abuse_number_of_messages_per_1h --value=30
 occ config:app:set mail abuse_number_of_messages_per_1d --value=100
 ```
 
+## Google OAuth
+
+This app can allow users to connect their Google accounts with OAuth. This makes it possible to use accounts without 2FA or app password.
+
```
Comment (miaulalala) [inline doc/admin.md:74]: Should we mention app passwords as well? Or are they not going to work any longer?

**Review by miaulalala — APPROVED** [review]
Looks good to me, tested and works

**Review by GretaD — APPROVED** [review]
looks good

File: doc/admin.md (hunk 3 above)
Comment (ChristophWurst) [reply doc/admin.md:74]: They work, but only if you use 2FA.

**Pull request merged**